### PR TITLE
[Feature] Implement random distribution table

### PR DESCRIPTION
**Pull request by meegoo**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

### Why
* Simplify the complexity of creating table statements.
* Provide better out-of-box performance, avoid unbalanced distribution due to low cardinality distribution keys.
* Random distribution can use sequential data writing to improve ingestion performance.
### What
Random distribution is one of the ways to distribute data across buckets. 
By using a random distribution, StarRocks spreads the data evenly across all buckets without considering the values in any specific column.
If create table, create materialized view, CTAS does `not specify distribution`, use random distribution as `default`
#### Random Distribution Table/MaterializedView
```
# automatically determine buckets num 
CREATE TABLE t (
    k BIGINT,
    v BIGINT
) DUPLICATE KEY (k);

# specify buckets num
CREATE TABLE t (
    k BIGINT,
    v BIGINT
) DUPLICATE KEY (v)
DISTRIBUTED BY RANDOM BUCKETS 8;

# materialized view
create materialized view tv REFRESH ASYNC as select k, sum(v) from t group by k;
create materialized view tv distributed by random REFRESH ASYNC as select k, sum(v) from t group by k;
create materialized view tv distributed by random buckets 10 REFRESH ASYNC as select k, sum(v) from t group by k;

# CTAS
create table c as select * from t;
create table c distributed by random as select * from t;
create table c distributed by random buckets 10 as select * from t;
```
### Limitation
- Random distribution does not support PrimaryKey、UniqueKey、AggregationKey.
- Random distribution does not support colocate tables.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exec/tablet_info.cpp [hunk 1]
```diff
@@ -492,6 +492,13 @@ void OlapTablePartitionParam::_compute_hashes(Chunk* chunk, std::vector<uint32_t
         _distributed_columns[i] = chunk->get_column_by_slot_id(_distributed_slot_descs[i]->id()).get();
         _distributed_columns[i]->crc32_hash(&(*indexes)[0], 0, num_rows);
     }
+
+    // if no distributed columns, use random number
+    if (_distributed_slot_descs.size() == 0) {
+        for (auto i = 0; i < num_rows; ++i) {
+            (*indexes)[i] = _rand.Next();
```
Comment (imay) [inline be/src/exec/tablet_info.cpp:499]: Did you test its performance?
Why not just choose a first value and just increment it.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 2]
```diff
@@ -2046,6 +2046,7 @@ distributionClause
 distributionDesc
     : DISTRIBUTED BY HASH identifierList (BUCKETS INTEGER_VALUE)?
     | DISTRIBUTED BY HASH identifierList
+    | DISTRIBUTED BY RANDOM (BUCKETS INTEGER_VALUE)?
```
[the remark this answers is not in the document]
Comment (meegoo) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2049]: default distribution is random.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 32 (46.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|0|1|00.00%|[294]|
|:large_blue_circle:|com/starrocks/sql/ast/DistributionDesc.java|0|4|00.00%|[64, 65, 66, 67]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|6|16.67%|[1035, 1036, 1037, 1039, 2112]|
|:large_blue_circle:|com/starrocks/sql/ast/RandomDistributionDesc.java|1|2|50.00%|[71]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|8|14|57.14%|[461, 462, 463, 464, 465, 466]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|1|1|100.00%|[]|

**Review by jaogoy — COMMENTED** [review]
Will it influence the query bucket pruning?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 18 / 35 (51.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/DistributionDesc.java|0|4|00.00%|[64, 65, 66, 67]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|6|16.67%|[1035, 1036, 1037, 1039, 2112]|
|:large_blue_circle:|com/starrocks/sql/ast/RandomDistributionDesc.java|1|2|50.00%|[71]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|8|14|57.14%|[461, 462, 463, 464, 465, 466]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|2|3|66.67%|[344]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 60 / 78 (76.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|0|1|00.00%|[304]|
|:large_blue_circle:|com/starrocks/sql/ast/DistributionDesc.java|0|4|00.00%|[64, 65, 66, 67]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|10|40.00%|[1035, 1036, 1037, 1039, 1636, 1637]|
|:large_blue_circle:|com/starrocks/sql/ast/RandomDistributionDesc.java|1|2|50.00%|[71]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|2|3|66.67%|[353]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|4|75.00%|[172]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|3|4|75.00%|[125]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|4|5|80.00%|[334]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|11|13|84.62%|[465, 467]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|

File: be/src/exec/tablet_info.cpp [hunk 3]
```diff
@@ -492,6 +492,14 @@ void OlapTablePartitionParam::_compute_hashes(Chunk* chunk, std::vector<uint32_t
         _distributed_columns[i] = chunk->get_column_by_slot_id(_distributed_slot_descs[i]->id()).get();
         _distributed_columns[i]->crc32_hash(&(*indexes)[0], 0, num_rows);
     }
+
+    // if no distributed columns, use random distribution
+    if (_distributed_slot_descs.size() == 0) {
```
Comment (chaoyli) [inline be/src/exec/tablet_info.cpp:497]: What's the performance improvement after enabling this feature?

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/DistributionDesc.java [hunk 4]
```diff
@@ -61,6 +61,10 @@ public static DistributionDesc read(DataInput in) throws IOException {
             DistributionDesc desc = new HashDistributionDesc();
             desc.readFields(in);
             return desc;
+        } else if (type == DistributionInfoType.RANDOM) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/DistributionDesc.java:64]: Do we need support to convert the hash distribution to random distribution?

**Comment by chaoyli** [comment]
@meegoo Do you think? Which of Random Distribution or Even Distribution is better?

File: be/src/exec/tablet_info.cpp (hunk 3 above)
Comment (meegoo) [reply be/src/exec/tablet_info.cpp:497]: non improve

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/DistributionDesc.java (hunk 4 above)
Comment (meegoo) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/DistributionDesc.java:64]: using CTAS now

**Comment by mofeiatwork** [comment]
does it work for CTAS ?

**Comment by meegoo** [comment]
> does it work for CTAS ?

yes

> does it work for CTAS ?

yes

**Review by chaoyli — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 61 / 90 (67.78%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|7|00.00%|[1034, 1035, 1036, 1038, 1635, 1636, 2116]|
|:large_blue_circle:|com/starrocks/sql/ast/DistributionDesc.java|0|4|00.00%|[64, 65, 66, 67]|
|:large_blue_circle:|com/starrocks/sql/ast/RandomDistributionDesc.java|1|2|50.00%|[71]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CTASAnalyzer.java|11|18|61.11%|[149, 150, 151, 152, 153, 154, 155]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|2|3|66.67%|[365]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|9|66.67%|[722, 723, 724]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|4|75.00%|[172]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|3|4|75.00%|[125]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|4|5|80.00%|[334]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|6|83.33%|[5932]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|11|13|84.62%|[465, 467]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 73 / 97 (75.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/DistributionDesc.java|0|4|00.00%|[64, 65, 66, 67]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|10|40.00%|[1034, 1035, 1036, 1038, 1635, 1636]|
|:large_blue_circle:|com/starrocks/sql/ast/RandomDistributionDesc.java|1|2|50.00%|[71]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CTASAnalyzer.java|12|19|63.16%|[150, 151, 152, 153, 154, 155, 156]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|2|3|66.67%|[365]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|4|75.00%|[172]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|3|4|75.00%|[125]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|4|5|80.00%|[334]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|11|13|84.62%|[465, 467]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|

**Review by chaoyli — DISMISSED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**